src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2352,51 +2434,105 @@ private void ReportNullabilityMismatchWithTargetDelegate(SyntaxNode syntax, Name
                 case ConversionKind.ImplicitUserDefined:
-                    var methodOpt = conversion.Method;
-                    if ((object)methodOpt != null && methodOpt.ParameterCount == 1)
+                    // cf. Binder.CreateUserDefinedConversion
                     {
-                        if (conversion.BestUserDefinedConversionAnalysis.Kind == UserDefinedConversionAnalysisKind.ApplicableInLiftedForm)
+                        if (!conversion.IsValid)
                         {
-                            // PROTOTYPE(NullableReferenceTypes): Conversions: Lifted operator
+                            canConvert = false;
                             break;
                         }
-                        // PROTOTYPE(NullableReferenceTypes): Update method based on operandType.
-                        var resultType = methodOpt.ReturnType;
-                        // Ensure converted type matches expected target type, specifically for lifted
-                        // conversions. See LocalRewriter.MakeConversionNode for similar handling
-                        // based on conversion.BestUserDefinedConversionAnalysis.ToType.
-                        if (!fromConversionNode)
-                        {
-                            if (resultType.TypeSymbol != targetType)
-                            {
-                                // PROTOTYPE(NullableReferenceTypes): Should handle nested nullability
-                                // and top-level nullability. Conversion from resultType to targetType
-                                // should be treated as a built-in conversion.
-                                resultType = TypeSymbolWithAnnotations.Create(targetType);
-                            }
-                        }
-                        return resultType;
+
+                        // operand -> conversion "from" type
+                        operandType = ApplyConversion(
+                            node,
+                            operandOpt,
+                            conversion.UserDefinedFromConversion,
+                            conversion.BestUserDefinedConversionAnalysis.FromType,
+                            operandType,
+                            checkConversion: false,
+                            fromExplicitCast: false,
+                            out _);
+
+                        // PROTOTYPE(NullableReferenceTypes): Update method based on operandType
+                        // (see StaticNullChecking_FlowAnalysis.Conversions_07).
+                        var methodOpt = conversion.Method;
+                        Debug.Assert((object)methodOpt != null);
+                        Debug.Assert(methodOpt.ParameterCount == 1);
+                        var parameter = methodOpt.Parameters[0];
+
+                        // conversion "from" type -> method parameter type
+                        operandType = ClassifyAndApplyConversion(node, parameter.Type.TypeSymbol, operandType);
```

Comment:
>operandType = ClassifyAndApplyConversion(node, parameter.Type.TypeSymbol, operandType); [](start = 24, length = 87)

Is this responsible for detecting top level mismatch?

Reply:
No, the following method is used for reporting top-level nullability mismatch. Changed the following method to an explicit call to `ReportNullReferenceArgumentIfNecessary` to make that more obvious.

---
In reply to: [190745125](https://github.com/dotnet/roslyn/pull/26356#discussion_r190745125) [](ancestors = 190745125)